dali/kernels/dynamic_scratchpad_test.cc

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/kernels/dynamic_scratchpad.h"  // NOLINT
+#include <gtest/gtest.h>
+#include <algorithm>
+#include <iostream>
+#include <chrono>
+#include <random>
+#include <vector>
+#include "dali/core/cuda_utils.h"
+#include "dali/core/cuda_stream_pool.h"
+#include "dali/core/mm/memory.h"
+
+namespace dali {
+namespace kernels {
+namespace test {
+
+/**
+ * @brief Tests basic dynamic scratchpad functioning
+ *
+ * This test checks that:
+ * - the memory is usable and accessible on the right backend
+ * - the pinned memory block is released in stream order, which prevents
+ *   immediate reuse if the stream is still running
+ * - it makes multiple attempts to catch the stream still running
+ */
+TEST(DynamicScratchpad, BasicTest) {
+  const int N = 64 << 10;  // 64 KiB
+
+  std::vector<char> in(N);
+  for (int i = 0; i < N; i++)
+    in[i] = i + 42;  // so it doesn't start or end with 0
+
+  auto stream = CUDAStreamPool::instance().Get();
+  auto dev = mm::alloc_raw_unique<char, mm::memory_kind::device>(N);
+  int max_attempts = 1000;
+  bool was_running = false;
+  for (int attempt = 0; attempt < max_attempts; attempt++) {
+    char *pinned;
+    {
+      DynamicScratchpad scratch({}, AccessOrder(stream));
+      pinned = scratch.Allocate<mm::memory_kind::pinned, char>(N);
+      memcpy(pinned, in.data(), N);
+      CUDA_CALL(cudaMemcpyAsync(dev.get(), pinned, N, cudaMemcpyHostToDevice, stream));
+    }
+    auto out = mm::alloc_raw_unique<char, mm::memory_kind::pinned>(N);
+    bool running = false;
+    if (was_running) {
+      CUDA_CALL(cudaStreamSynchronize(stream));
+    } else {
+      running = cudaStreamQuery(stream) == cudaErrorNotReady;
+      if (running)
+        was_running = true;
+    }
+    ASSERT_TRUE(out.get() + N < pinned || out.get() >= pinned + N || !running);
+    CUDA_CALL(cudaMemcpyAsync(out.get(), dev.get(), N, cudaMemcpyDeviceToHost, stream));
+    CUDA_CALL(cudaStreamSynchronize(stream));
+    ASSERT_EQ(memcmp(in.data(), out.get(), N), 0);
+    if (was_running && !running)
+      break;
+  }
+  if (!was_running)
```

Comment:
how about `ASSERT_TRUE(was_running)`?

Reply:
I don't want it to just fail. On some machines it might be impossible to reach this kind of concurrency, e.g. due to CPU load.